store/tikv/region_request_test.go

```diff
@@ -569,3 +572,129 @@ func (s *testRegionRequestToSingleStoreSuite) TestOnMaxTimestampNotSyncedError(c
 		c.Assert(resp, NotNil)
 	}()
 }
+
+func (s *testRegionRequestToThreeStoresSuite) loadAndGetLeaderStore(c *C) (*Store, string) {
+	region, err := s.regionRequestSender.regionCache.findRegionByKey(s.bo, []byte("a"), false)
+	c.Assert(err, IsNil)
+	leaderStore, leaderPeer, _, leaderStoreIdx := region.WorkStorePeer(region.getStore())
+	c.Assert(leaderPeer.Id, Equals, s.leaderPeer)
+	leaderAddr, err := s.regionRequestSender.regionCache.getStoreAddr(s.bo, region, leaderStore, leaderStoreIdx)
+	c.Assert(err, IsNil)
+	return leaderStore, leaderAddr
+}
+
+func (s *testRegionRequestToThreeStoresSuite) TestForwarding(c *C) {
+	EnableForwarding = true
+	defer func() { EnableForwarding = false }()
+
+	// First get the leader's addr from region cache
+	leaderStore, leaderAddr := s.loadAndGetLeaderStore(c)
+
+	bo := NewBackoffer(context.Background(), 10000)
+
+	// Simulate that the leader is network-partitioned but can be accessed by forwarding via a follower
+	innerClient := s.regionRequestSender.client
+	s.regionRequestSender.client = &fnClient{fn: func(ctx context.Context, addr string, req *tikvrpc.Request, timeout time.Duration) (*tikvrpc.Response, error) {
+		if addr == leaderAddr {
+			return nil, errors.New("simulated rpc error")
+		}
+		// MockTiKV doesn't support forwarding. Simulate forwarding here.
+		if len(req.ForwardedHost) != 0 {
+			addr = req.ForwardedHost
+		}
+		return innerClient.SendRequest(ctx, addr, req, timeout)
+	}}
+
+	loc, err := s.regionRequestSender.regionCache.LocateKey(bo, []byte("k"))
+	c.Assert(err, IsNil)
+	c.Assert(loc.Region.GetID(), Equals, s.regionID)
+	req := tikvrpc.NewRequest(tikvrpc.CmdRawPut, &kvrpcpb.RawPutRequest{
+		Key:   []byte("k"),
+		Value: []byte("v1"),
+	})
+	resp, ctx, err := s.regionRequestSender.SendReqCtx(bo, req, loc.Region, time.Second, kv.TiKV)
```

Comment:
Trigger request forwarding because `mocktikv` doesn't implement the health service? I prefer to use the failpoint here.
```golang
	c.Assert(failpoint.Enable("github.com/pingcap/tidb/store/tikv/mockRequestLiveness", "return(false)"), IsNil)
```

Reply:
It's hard to let the failpoint know what address to refuse, and whether forwarded reqeusts should success, to do this we need to pass there information with a string, and parse it in the failpoint. So I think the `fnClient` (which is already provided) is more easy to use.

Reply:
And `mockRequestLiveness` only injects the `requestLiveness` function, which won't be executed if no RPC error occurs.

Reply:
My point is when `fnClient` reports error, the `requestLiveness` fails because the `mocktikv` doesn't implement health service now. I prefer to use the fail point to make `requestLiveness` fail.

Reply:
Never mind. The fail point is removed. 🤣